File: backend/models/RoomMember.js
const mongoose = require('mongoose');

/**
 * RoomMember Schema - Stores Encrypted Session Keys per Member
 * 
 * Critical Security Component:
 * Each member has their own encrypted copy of the room's SessionKey.
 * 
 * Key Distribution Process:
 * 1. Admin generates SessionKey (256-bit AES key)
 * 2. For each member, encrypt SessionKey with their RSA public key:
 *    encryptedSessionKey = RSA_Encrypt(sessionKey, memberPublicKey)
 * 3. Store each encrypted copy in this collection
 * 4. When member joins, send them their encrypted copy
 * 5. Member decrypts: sessionKey = RSA_Decrypt(encryptedSessionKey, memberPrivateKey)
 * 
 * Key Rotation (after kicking member):
 * 1. Admin generates NEW SessionKey
 * 2. Delete kicked member's RoomMember entry
 * 3. For remaining members, encrypt new SessionKey and update encryptedSessionKey
 * 4. Increment room's sessionKeyVersion
 * 5. Kicked member can't decrypt new messages (has old key)
 */
const roomMemberSchema = new mongoose.Schema({
  // References
  roomId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Room',
    required: true
  },
  
  userId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  
  // E2EE Key Storage
  encryptedSessionKey: {
    type: String,
    required: true,
    // This is the SessionKey encrypted with the member's PUBLIC key
    // Format: Base64-encoded RSA-OAEP encrypted data
    // Only this specific user can decrypt it with their PRIVATE key
  },
  
  sessionKeyVersion: {
    type: Number,
    required: true,
    default: 1,
    // Must match Room.sessionKeyVersion
    // If user's version < room's version, they need to fetch new key
  },
  
  // Member role and permissions
  role: {
    type: String,
    enum: ['owner', 'admin', 'member'],
    default: 'member'
  },
  
  // Member metadata
  joinedAt: {
    type: Date,
    default: Date.now
  },
  
  lastSeenMessageId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Message',
    default: null
  },
  
  unreadCount: {
    type: Number,
    default: 0
  },
  
  // Notification settings
  isMuted: {
    type: Boolean,
    default: false
  },
  
  mutedUntil: {
    type: Date,
    default: null
  },
  
  // Status
  isActive: {
    type: Boolean,
    default: true
    // False if kicked/left. Keep record for history but don't send messages
  },
  
  leftAt: {
    type: Date,
    default: null
  }
});

// Compound indexes for efficient queries
roomMemberSchema.index({ roomId: 1, userId: 1 }, { unique: true });
roomMemberSchema.index({ userId: 1, isActive: 1 });
roomMemberSchema.index({ roomId: 1, isActive: 1 });

// Methods
roomMemberSchema.methods = {
  // Check if member has latest key version
  hasLatestKey(roomKeyVersion) {
    return this.sessionKeyVersion === roomKeyVersion;
  },
  
  // Update session key (during key rotation)
  async updateSessionKey(newEncryptedKey, newVersion) {
    this.encryptedSessionKey = newEncryptedKey;
    this.sessionKeyVersion = newVersion;
    await this.save();
  },
  
  // Mark as left/kicked
  async deactivate() {
    this.isActive = false;
    this.leftAt = Date.now();
    await this.save();
  }
};

// Statics
roomMemberSchema.statics = {
  // Get member's encrypted key for a room
  async getMemberKey(roomId, userId) {
    return this.findOne({ 
      roomId, 
      userId, 
      isActive: true 
    });
  },
  
  // Get all active members of a room with their keys
  async getRoomMembers(roomId) {
    return this.find({ roomId, isActive: true })
      .populate('userId', 'username email avatar publicKey')
      .sort({ joinedAt: 1 });
  },
  
  // Get rooms for a user
  async getUserRooms(userId) {
    return this.find({ userId, isActive: true })
      .populate({
        path: 'roomId',
        populate: {
          path: 'ownerId',
          select: 'username email avatar'
        }
      })
      .sort({ 'roomId.lastMessageAt': -1 });
  },
  
  // Bulk create member entries (when creating room or adding multiple members)
  async createMemberEntries(entries) {
    // entries = [{ roomId, userId, encryptedSessionKey, sessionKeyVersion, role }]
    return this.insertMany(entries);
  },
  
  // Delete member entry (when kicked or left)
  async removeMember(roomId, userId) {
    return this.findOneAndUpdate(
      { roomId, userId },
      { isActive: false, leftAt: Date.now() }
    );
  }
};

module.exports = mongoose.model('RoomMember', roomMemberSchema);

File: flutter/lib/models/room_member.dart
/**
 * RoomMember Model - Represents a member of a group chat room
 * 
 * Critical Security Component:
 * - Each member has their own encrypted copy of the SessionKey
 * - encryptedSessionKey: Encrypted with member's RSA public key
 * - Only this member can decrypt it with their private key
 */
class RoomMember {
  final String id;
  final String roomId;
  final String userId;
  final String encryptedSessionKey; // Base64 encoded RSA-encrypted session key
  final int sessionKeyVersion;
  final String role; // 'owner', 'admin', 'member'
  final DateTime joinedAt;
  final int unreadCount;
  final bool isMuted;
  final DateTime? mutedUntil;
  final bool isActive;

  // User info (populated from join)
  final String? username;
  final String? email;
  final String? avatar;

  RoomMember({
    required this.id,
    required this.roomId,
    required this.userId,
    required this.encryptedSessionKey,
    required this.sessionKeyVersion,
    required this.role,
    required this.joinedAt,
    this.unreadCount = 0,
    this.isMuted = false,
    this.mutedUntil,
    this.isActive = true,
    this.username,
    this.email,
    this.avatar,
  });

  factory RoomMember.fromJson(Map<String, dynamic> json) {
    return RoomMember(
      id: json['id'] ?? json['_id'] ?? '',
      roomId: json['roomId'] ?? '',
      userId: json['userId'] ?? json['user']?['id'] ?? json['user']?['_id'] ?? '',
      encryptedSessionKey: json['encryptedSessionKey'] ?? '',
      sessionKeyVersion: json['sessionKeyVersion'] ?? 1,
      role: json['role'] ?? 'member',
      joinedAt: json['joinedAt'] != null
          ? DateTime.parse(json['joinedAt'])
          : DateTime.now(),
      unreadCount: json['unreadCount'] ?? 0,
      isMuted: json['isMuted'] ?? false,
      mutedUntil: json['mutedUntil'] != null
          ? DateTime.parse(json['mutedUntil'])
          : null,
      isActive: json['isActive'] ?? true,
      username: json['username'] ?? json['user']?['username'],
      email: json['email'] ?? json['user']?['email'],
      avatar: json['avatar'] ?? json['user']?['avatar'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'roomId': roomId,
      'userId': userId,
      'encryptedSessionKey': encryptedSessionKey,
      'sessionKeyVersion': sessionKeyVersion,
      'role': role,
      'joinedAt': joinedAt.toIso8601String(),
      'unreadCount': unreadCount,
      'isMuted': isMuted,
      'mutedUntil': mutedUntil?.toIso8601String(),
      'isActive': isActive,
      'username': username,
      'email': email,
      'avatar': avatar,
    };
  }

  // SQLite database methods
  factory RoomMember.fromDatabase(Map<String, dynamic> row) {
    return RoomMember(
      id: row['id'] as String,
      roomId: row['room_id'] as String,
      userId: row['user_id'] as String,
      encryptedSessionKey: row['encrypted_session_key'] as String,
      sessionKeyVersion: row['session_key_version'] as int,
      role: row['role'] as String,
      joinedAt: DateTime.parse(row['joined_at'] as String),
      unreadCount: row['unread_count'] as int? ?? 0,
      isMuted: (row['is_muted'] as int? ?? 0) == 1,
      mutedUntil: row['muted_until'] != null
          ? DateTime.parse(row['muted_until'] as String)
          : null,
      isActive: (row['is_active'] as int? ?? 1) == 1,
      username: row['username'] as String?,
      email: row['email'] as String?,
      avatar: row['avatar'] as String?,
    );
  }

  Map<String, dynamic> toDatabase() {
    return {
      'id': id,
      'room_id': roomId,
      'user_id': userId,
      'encrypted_session_key': encryptedSessionKey,
      'session_key_version': sessionKeyVersion,
      'role': role,
      'joined_at': joinedAt.toIso8601String(),
      'unread_count': unreadCount,
      'is_muted': isMuted ? 1 : 0,
      'muted_until': mutedUntil?.toIso8601String(),
      'is_active': isActive ? 1 : 0,
    };
  }

  // Check if member has latest key version
  bool hasLatestKey(int roomKeyVersion) {
    return sessionKeyVersion == roomKeyVersion;
  }

  // Create copy with updated fields
  RoomMember copyWith({
    String? encryptedSessionKey,
    int? sessionKeyVersion,
    int? unreadCount,
    bool? isMuted,
    DateTime? mutedUntil,
  }) {
    return RoomMember(
      id: id,
      roomId: roomId,
      userId: userId,
      encryptedSessionKey: encryptedSessionKey ?? this.encryptedSessionKey,
      sessionKeyVersion: sessionKeyVersion ?? this.sessionKeyVersion,
      role: role,
      joinedAt: joinedAt,
      unreadCount: unreadCount ?? this.unreadCount,
      isMuted: isMuted ?? this.isMuted,
      mutedUntil: mutedUntil ?? this.mutedUntil,
      isActive: isActive,
      username: username,
      email: email,
      avatar: avatar,
    );
  }
}
